### ISSUE TITLE
[info]  RPM package

### ISSUE THREAD

**Issue by remicollet** [DataDog/dd-trace-php#214]
For your information

RPM packages for Fedora / RHEL / CentOS are already available, among some other extensions, in my repository

https://blog.remirepo.net/pages/PECL-extensions-RPM-status

On Fedora (and RHEL-8-beta)

`dnf install php-pecl-datadog-trace`

On RHEL / CentOS 6 and 7

`yum install php-pecl-datadog-trace`

Notice: `yum install php-ddtrace` also works (using extension name)

Available for PHP 5.6, 7.0, 7.1 and 7.2.

Spec file tracked on 
https://git.remirepo.net/cgit/rpms/php/pecl/php-pecl-datadog-trace.git/

**Comment by SammyK** [comment]
Wow! That was fast - thanks so much @remicollet! :D

**Comment by pawelchcki** [comment]
Thanks for the headsup and great work on this @remicollet!

**Comment by JanThiel** [comment]
Hi @remicollet Thanks alot for your amazing work :-)
Did you ever had a chance to test if the tracer is actually working? I installed it for php73-php-fpm on CentOS 7 and am not able to get any traces out of it although it seems to be configured correctly. The DataDog support is also already investigating. Thus I thought maybe there is something you know or stumbled upon :-)

**Comment by JanThiel** [comment]
Just got an answer from the DataDog support which might be interesting for others looking for a way to get it running with the Remi php-safe packages:

`It seems we do not actually officially support the repo from Remi being used.`

Sad to hear this. So the answer might be simply, that it *does not* work at all with the `php73-php-ddtrace` package. Finally asking myself, if the package shouldn't be simply droped from the php-safe packages from remi.

**Comment by morrisonlevi** [comment]
@remicollet Are you basically packaging something that wraps `pecl install`?

**Comment by remicollet** [comment]
> @remicollet Are you basically packaging something that wraps pecl install?

Don't really understand the question,

The RPM provides the extension, installed in the proper location, and its configuration file so it is loaded by default and detected by PHP.

And also registered in pecl registry (if PEAR installed on the system)

More detail in the spec file https://git.remirepo.net/cgit/rpms/php/pecl/php-pecl-datadog-trace.git/tree/php-pecl-datadog-trace.spec  (mostly a shell script, quite easy to read).

**Comment by morrisonlevi** [comment]
@remicollet To get a fully-functioning install we need to also install the `bridge` and `src` folders somewhere. For pecl installs, I'm planning on making `datadog_trace/{src,bridge}` in PHP's libdir, usually something like `/usr/local/lib/php` or `/usr/local/php/lib`. Then, the ini setting `ddtrace.request_init_hook` needs to point at the installed `datadog_trace/bridge/dd_wrap_autoloader.php`.

It is unusual to ship PHP and C extensions simultaneously, but in our case the PHP files should be considered part of the extension, as they need to be kept in sync version by version.

**Comment by remicollet** [comment]
> Based on my reading of the spec file, that's all that is missing.

The spec file install the same stuff that "pecl instal" ;)

> It is unusual to ship PHP libraries and C extensions simultaneously

AFAIK, xhprof is the only one I know which does it

> For pecl installs, I'm planning on making 

As soon as new files will appear in the pecl archive, I will add them ;)

**Comment by swalberg** [comment]
Hi, wanted to echo that the `php74-php-pecl-datadog-trace` package from Remi doesn't work. I used the self test script at https://github.com/DataDog/dd-trace-php/blob/master/src/dd-doctor.php and the `ddtrace version (in userland)` shows as an empty string.

If I add the `ddtrace.request_init_hook` to point to the `dd_wrap_autoloader.php` from the official Datadog distribution, it passes except for `Integrations loaded`.

**Comment by remicollet** [comment]
@swalberg everything seems OK.

```
$ DD_TRACE_CLI_ENABLED=1 php -n -d extension=curl -d extension=ddtrace -d ddtrace.request_init_hook=$PWD/bridge/dd_wrap_autoloader.php src/dd-doctor.php 
WARNING: Script is running from the CLI SAPI.
         Please run this script from your web browser.

DataDog trace extension verification

- PHP version and SAPI                               [7.4.5 - cli]
- ddtrace extension installed                        [OK]
- ddtrace version (installed)                        [0.43.0]
- ddtrace version (const)                            [0.43.0]
- ddtrace version (userland)                         [0.43.0]
- ddtrace versions in sync                           [OK]
- dd_trace() function available                      [OK]
- dd_trace_env_config() function available           [OK]
- ddtrace.request_init_hook set                      [OK]
- ddtrace.request_init_hook reachable                [OK]
- ddtrace.request_init_hook has run                  [OK]
- 'open_basedir' allows access to '/proc/self/'      [OK]
- IntegrationsLoader exists                          [OK]
- Integrations not loaded yet                        [OK]
- Registering an autoloader...
- Integrations loaded                                [OK]
- DDTrace\Tracer class exists                        [OK]
- Background sender is enabled?                      [NO]
  > You can enable the background sender via
    DD_TRACE_BETA_SEND_TRACES_VIA_THREAD=true
- Configured Agent host                              [localhost]
- Configured Agent port                              [8126]
- Agent can receive traces                           [FAIL]
Curl verbose output: 

*   Trying 127.0.0.1:8126...
* TCP_NODELAY set
* connect to 127.0.0.1 port 8126 failed: Connection refused
* Failed to connect to localhost port 8126: Connection refused
* Closing connection 0
```

**Comment by swalberg** [comment]
Hi Remi, thanks for responding.

I noticed that in your command you set `ddtrace.request_init_hook` and have the Datadog wrapper in your directory. When I installed `php74-php-pecl-datadog-trace` it did not come with the wrapper. So while the package does technically install the Datadog agent, it's done in such a way that the agent doesn't do anything useful. For reference, the official Datadog package includes both the agent and the wrapper, as discussed above.

I tried copying those directories from another server with the 0.41.0 agent but it seems that the wrapper version needs to be in sync with the agent version. So at least for me, this `php74-php-pecl-datadog-trace` isn't enough.

I fear we're getting into a game of semantics here where "OK" to you means that the agent is loaded into PHP, but I'm calling it "OK" once it instruments apps out of the box like the Datadog official package does. Would you agree?

**Comment by swalberg** [comment]
Just to split the middle on this, I found that the official datadog RPM works with the Remi distro of PHP 7.4 if you first make `php` available in the search path:

`ln -s /opt/remi/php74/root/usr/bin/php /usr/local/bin/php`



```Running transaction
  Installing : datadog-php-tracer-0.41.1-1.x86_64                                             1/1
### Installing Datadog PHP tracing extension (ddtrace)
###
### Logging /usr/local/bin/php -i to a file
###
Cannot load Zend OPcache - it was already loaded
Cannot load Zend OPcache - it was already loaded
Cannot load Zend OPcache - it was already loaded
Cannot load Zend OPcache - it was already loaded
### Creating ddtrace.ini
###

; ***** DO NOT EDIT THIS FILE *****
; To overwrite the INI settings for this extension, edit
; the INI file in this directory called "ddtrace-custom.ini"

[datadog]
extension=/opt/datadog-php/extensions/ddtrace-20190902.so
ddtrace.request_init_hook=/opt/datadog-php/dd-trace-sources/bridge/dd_wrap_autoloader.php
### ddtrace.ini created
###
### Created empty /opt/datadog-php/etc/ddtrace-custom.ini
###
### Linking ddtrace.ini for supported SAPI's
###

### Found SAPI config directory: /etc/opt/remi/php74/php.d
### Linking ddtrace.ini to /etc/opt/remi/php74/php.d/98-ddtrace.ini
### Linking ddtrace-custom.ini to /etc/opt/remi/php74/php.d/99-ddtrace-custom.ini
###
Cannot load Zend OPcache - it was already loaded
ddtrace
### Extension enabled successfully
  Verifying  : datadog-php-tracer-0.41.1-1.x86_64                                             1/1

Installed:
  datadog-php-tracer.x86_64 0:0.41.1-1

Complete!```

**Comment by remicollet** [comment]
> I fear we're getting into a game of semantics here where "OK" to you means that the agent is loaded into PHP, but I'm calling it "OK" once it instruments apps out of the box like the Datadog official package does. Would you agree?

I mean than "yum instal ..."  has the same result than "pecl instal ..." (without the need of having to build anything, not having to install development tools, which is a bas practice on a prod server)

**Comment by remicollet** [comment]
Out of topic here, but:

> Just to split the middle on this, I found that the official datadog RPM works with the Remi distro of PHP 7.4 if you first make `php` available in the search path:
> 
> `ln -s /opt/remi/php74/root/usr/bin/php /usr/local/bin/php`

Which mean you are not properly using my repo ;)

If you want to have 7.4 in the search path, simply use "module load php74"
Or if you need a single version, use the proper package (`php-*` instead of `php74-php-*`) as explained in the [FAQ](https://blog.remirepo.net/pages/English-FAQ#scl) and by the [Wizard](https://rpms.remirepo.net/wizard/).

**Comment by remicollet** [comment]
See pr #845 which can fix the pecl installation issue, and then the RPM one.

**Comment by swalberg** [comment]
Hey Remi, thanks for the advice on how to use the repo. I only want one version of PHP so replacing system packages is fine. I'll look into that on Monday.

**Comment by remicollet** [comment]
So, with **0.48.0**, the RPM package now also provides the Library, and the proper configuration to use it (as for pecl package)

**pecl installed files**

```
$ pecl list-files datadog_trace
Installed Files For datadog_trace
=================================
Type Install Path
doc  /usr/share/doc/pecl/datadog_trace/src/ext/mpack/AUTHORS.md
doc  /usr/share/doc/pecl/datadog_trace/src/ext/mpack/CHANGELOG.md
doc  /usr/share/doc/pecl/datadog_trace/src/ext/mpack/README.md
doc  /usr/share/doc/pecl/datadog_trace/CHANGELOG.md
doc  /usr/share/doc/pecl/datadog_trace/NOTICE
doc  /usr/share/doc/pecl/datadog_trace/README.md
doc  /usr/share/doc/pecl/datadog_trace/UPGRADE-0.10.md
php  /usr/share/pear/datadog_trace/bridge/_generated.php
php  /usr/share/pear/datadog_trace/bridge/autoload.php
php  /usr/share/pear/datadog_trace/bridge/dd_init.php
php  /usr/share/pear/datadog_trace/bridge/dd_register_optional_deps_autoloader.php
php  /usr/share/pear/datadog_trace/bridge/dd_wrap_autoloader.php
php  /usr/share/pear/datadog_trace/bridge/php5.php
```

**RPM installed files**

```
$ rpm -ql php-pecl-datadog-trace 
/etc/php-zts.d/40-ddtrace.ini
/etc/php.d/40-ddtrace.ini
/usr/lib/.build-id
/usr/lib/.build-id/be
/usr/lib/.build-id/be/8facaf3aae34d1a6ccf8acfda34fb95e73bea6
/usr/lib/.build-id/f3
/usr/lib/.build-id/f3/81077c492334492ba6a7c0826cded80f13b340
/usr/lib64/php-zts/modules/ddtrace.so
/usr/lib64/php/modules/ddtrace.so
/usr/share/doc/pecl/ddtrace
/usr/share/doc/pecl/ddtrace/CHANGELOG.md
/usr/share/doc/pecl/ddtrace/NOTICE
/usr/share/doc/pecl/ddtrace/README.md
/usr/share/doc/pecl/ddtrace/UPGRADE-0.10.md
/usr/share/doc/pecl/ddtrace/src
/usr/share/doc/pecl/ddtrace/src/ext
/usr/share/doc/pecl/ddtrace/src/ext/mpack
/usr/share/doc/pecl/ddtrace/src/ext/mpack/AUTHORS.md
/usr/share/doc/pecl/ddtrace/src/ext/mpack/CHANGELOG.md
/usr/share/doc/pecl/ddtrace/src/ext/mpack/README.md
/usr/share/licenses/php-pecl-datadog-trace
/usr/share/licenses/php-pecl-datadog-trace/LICENSE
/usr/share/licenses/php-pecl-datadog-trace/LICENSE.Apache
/usr/share/licenses/php-pecl-datadog-trace/LICENSE.BSD3
/usr/share/licenses/php-pecl-datadog-trace/LICENSE.mpack
/usr/share/pear/datadog_trace
/usr/share/pear/datadog_trace/bridge
/usr/share/pear/datadog_trace/bridge/_generated.php
/usr/share/pear/datadog_trace/bridge/autoload.php
/usr/share/pear/datadog_trace/bridge/dd_init.php
/usr/share/pear/datadog_trace/bridge/dd_register_optional_deps_autoloader.php
/usr/share/pear/datadog_trace/bridge/dd_wrap_autoloader.php
/usr/share/pear/datadog_trace/bridge/php5.php
/var/lib/php/peclxml/php-pecl-datadog-trace.xml
```

**Default configuration**

```
$ php  --ri ddtrace

ddtrace


Datadog PHP tracer extension
For help, check out the documentation at https://docs.datadoghq.com/tracing/languages/php/
(c) Datadog 2020

Datadog tracing support => enabled
Version => 0.48.0
DATADOG TRACER CONFIGURATION => {"date":"2020-08-26T07:31:45Z","os_name":"Linux builder.remirepo.net 5.7.15-100.fc31.x86_64 #1 SMP Tue Aug 11 17:18:01 UTC 2020 x86_64","os_version":"5.7.15-100.fc31.x86_64","version":"0.48.0","lang":"php","lang_version":"7.3.21","env":null,"enabled":true,"service":null,"enabled_cli":false,"agent_url":"http://localhost:8126","debug":false,"analytics_enabled":false,"sample_rate":1.000000,"sampling_rules":null,"tags":null,"service_mapping":null,"distributed_tracing_enabled":true,"priority_sampling_enabled":true,"dd_version":null,"architecture":"x86_64","sapi":"cli","ddtrace.request_init_hook":"/usr/share/pear/datadog_trace/bridge/dd_wrap_autoloader.php","open_basedir_configured":false,"uri_fragment_regex":null,"uri_mapping_incoming":null,"uri_mapping_outgoing":null,"auto_flush_enabled":false,"generate_root_span":true,"http_client_split_by_domain":false,"measure_compile_time":true,"report_hostname_on_root_span":false,"traced_internal_functions":null,"auto_prepend_file_configured":false,"integrations_disabled":null,"enabled_from_env":true,"opcache.file_cache":null,"agent_error":"Failed to connect to localhost port 8126: Connection refused"}

                               Diagnostics                               
agent_error => Failed to connect to localhost port 8126: Connection refused
Diagnostic checks => failed

Directive => Local Value => Master Value
ddtrace.disable => Off => Off
ddtrace.request_init_hook => /usr/share/pear/datadog_trace/bridge/dd_wrap_autoloader.php => /usr/share/pear/datadog_trace/bridge/dd_wrap_autoloader.php

```

**Comment by SammyK** [comment]
Closing this as all of the PECL/Remi repo issues should be resolved as of `0.48.0`. Thanks again for all the help with this @remicollet!
